test/alternator/test_metrics.py

```diff
@@ -208,6 +211,73 @@ def test_streams_operations(test_table_s, dynamodbstreams, metrics):
+    # TODO: Also look at the latency_bucket.
+    yield
+    the_metrics = get_metrics(metrics)
+    for op in operation_names:
+        # The total "count" and "sum" on all shards should strictly increase
+        assert saved_latency_count[op] < get_metric(metrics, 'scylla_alternator_op_latency_count', {'op': op}, the_metrics)
+        assert saved_latency_sum[op] < get_metric(metrics, 'scylla_alternator_op_latency_sum', {'op': op}, the_metrics)
+
+# Test latency metrics for PutItem, GetItem, DeleteItem, UpdateItem.
+# We can't check what exactly the latency is - just that it gets updated.
+def test_item_latency(test_table_s, metrics):
+    with check_sets_latency(metrics, ['DeleteItem', 'GetItem', 'PutItem', 'UpdateItem']):
+        p = random_string()
+        test_table_s.put_item(Item={'p': p})
+        test_table_s.get_item(Key={'p': p})
+        test_table_s.delete_item(Key={'p': p})
+        test_table_s.update_item(Key={'p': p})
+
+# Test latency metrics for GetRecords. Other Streams-related operations -
+# ListStreams, DescribeStream, and GetShardIterator, have an operation
+# count (tested above) but do NOT currently have a latency histogram.
+def test_streams_latency(dynamodb, dynamodbstreams, metrics):
+    # Whereas the *count* metric for various operations also counts failed
+    # operations so we could write this test without a real stream, the
+    # latency metrics are only updated for *successful* operations so we
+    # need to use a real Alternator Stream in this test.
+    with new_test_table(dynamodb,
+        # Alternator Streams is expected to fail with tablets due to #16317.
+        # To ensure that this test still runs, instead of xfailing it, we
+        # temporarily coerce Altenator to avoid using default tablets
+        # setting, even if it's available. We do this by using the following
+        # tags when creating the table:
+        Tags=[{'Key': 'experimental:initial_tablets', 'Value': 'none'}],
+        KeySchema=[{ 'AttributeName': 'p', 'KeyType': 'HASH' }],
+        AttributeDefinitions=[{ 'AttributeName': 'p', 'AttributeType': 'S' }],
+        StreamSpecification={ 'StreamEnabled': True, 'StreamViewType': 'NEW_AND_OLD_IMAGES'}
+        ) as table:
+        # Wait for the stream to become active. This should be instantenous
+        # in Alternator, so the following loop won't wait
+        stream_enabled = False
+        start_time = time.time()
+        while time.time() < start_time + 60:
+            desc = table.meta.client.describe_table(TableName=table.name)['Table']
+            if 'LatestStreamArn' in desc:
+                arn = desc['LatestStreamArn']
+                desc = dynamodbstreams.describe_stream(StreamArn=arn)
+                if desc['StreamDescription']['StreamStatus'] == 'ENABLED':
+                    stream_enabled = True
+                    break
+            time.sleep(1)
```

Comment:
I wonder why we don't have such a loop in utils already. It could be extracted out and named as `timedout/expiring_(condition_)check` and would check for some condition for some amount of time, fail the test if the condition is not met.

Reply:
Maybe, but I don't think that such generic functions help anything - I think they only make things worse. These functions tend to accumulate more and more ugly parameters over time, and it becomes impossible to decide what defaults they should have. In this case, the function calls will replace just three lines (the setting of start_time, while and sleep) but in "return", it will complicate everything like how to pass or fail the test.

Maybe in this case, what could have been useful might have been a "wait for stream" function.  We actually have one in test_streams.py, called wait_for_active_stream(), but to be honest, it's a bit ugly and does things we don't need in this use case.

Since this test file is not about streams, and doesn't repeat this code, I really think that open-coding the three-line timeout loop plus the 5 lines of stream_enabled check, isn't so bad.